File: oc/Quartz2DDemo/Quartz2DDemo/PieView.m
//
//  PieView.m
//  Quartz2DDemo
//
//  Created by dfang on 2018-2-1.
//  Copyright © 2018年 east. All rights reserved.
//

#import "PieView.h"

@implementation PieView

- (void)drawRect:(CGRect)rect {
    float sum = 0;
    for (int i = 0; i < self.sections.count; i++) {
        sum = sum + [self.sections[i] floatValue];
    }

    CGFloat startAngle = -M_PI_2;
    CGFloat endAngle = 0;
    CGFloat radius = MIN(rect.size.width, rect.size.height) * 0.5;
    CGContextRef context = UIGraphicsGetCurrentContext();
    for (int i = 0; i < self.sectionColors.count; i++) {
        endAngle = [self.sections[i] floatValue] / sum * 2 * M_PI + startAngle;
        UIColor *curColor = self.sectionColors[i];
        [curColor set];
        CGContextMoveToPoint(context, rect.size.width * 0.5, rect.size.height * 0.5);
        CGContextAddArc(context, rect.size.width * 0.5, rect.size.height * 0.5, radius, startAngle, endAngle, 0);
        CGContextFillPath(context);
        startAngle = endAngle;
    }
}

@end

File: swift/Quartz2DDemo/Quartz2DDemo/PieView.swift
//
//  PieView.swift
//  Quartz2DDemo
//
//  Created by dfang on 2018-2-3.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class PieView: UIView {
    var sections:[Int] = []
    var sectionColors:[UIColor] = []
    
    override func draw(_ rect: CGRect) {
        var sum = 0
        
        for section in sections {
            sum += section
        }
        
        let context = UIGraphicsGetCurrentContext()
        var startAngle = -CGFloat(Float.pi/4)
        var endAngle:CGFloat = 0
        let center = CGPoint.init(x: rect.width * 0.5, y: rect.height * 0.5)
        let radius = min(rect.width, rect.height) * 0.5
        
        for color in sectionColors {
            endAngle = CGFloat(sections[sectionColors.index(of: color)!]) / CGFloat(sum) * 2.0 * CGFloat(Float.pi) + startAngle
            context?.setFillColor(color.cgColor)
            context?.move(to: center)
            context?.addArc(center: center, radius: radius, startAngle: startAngle, endAngle: endAngle, clockwise: false)
            context?.fillPath()
            startAngle = endAngle
        }
    }
    
}
